Emit used rustc invocation in the save-analysis file

Xanewok opened the pull request:
Blocked on https://github.com/nrc/rls-data/pull/19. (I'm guessing it won't pass CI due to an out-of-tree git dependency)

This should allow RLS to recreate a Rust compilation build plan from the save-analysis files alone, which should be useful when fetching those from external build systems, most notably Buck now.

Also this includes some more potentially useful compilation-specific options (e.g. sysroot or the actual path to extern crates) but that's not required for the build plan bits.

cc @jsgf @alexcrichton 

r? @nrc

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/430536361) ([raw log](https://api.travis-ci.org/v3/job/430536361/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
######################################################################    98.0%
#######################################################################   99.1%
######################################################################## 100.0%
[00:25:24] extracting /checkout/obj/build/cache/2018-09-11/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:25:24]     Updating git repository `https://github.com/Xanewok/rls-data`
[00:26:18]  Downloading filetime v0.2.1
[00:26:19]  Downloading cc v1.0.22
[00:26:19]  Downloading toml v0.4.6
[00:26:19]  Downloading lazy_static v0.2.11
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Xanewok commented:
One thing I was wondering was the current directory - should we include it in the save-analysis for the invocation or maybe we can assume that everything is relative to the directory of the save analysis file itself and just hope for the user to set it properly using `--remap-path-prefix`?

Xanewok commented:
Including crate_name from session options might not be a good idea, since it can be inferred and we include it in the `CrateId` anyways. What's interesting is that target_triple is inferred and filled in the session options, but crate_name or sysroot seem not to. At a first glance it seems inconsistent? But maybe there's more to it, I'll have to read further.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/431975905) ([raw log](https://api.travis-ci.org/v3/job/431975905/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
#######################################                                   54.7%
###################################################                       71.9%
######################################################################## 100.0%
[00:02:01] extracting /checkout/obj/build/cache/2018-09-11/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:02:01]     Updating git repository `https://github.com/Xanewok/rls-data`
[00:02:36]  Downloading serde v1.0.75
[00:02:37]  Downloading num_cpus v1.8.0
[00:02:37]  Downloading libc v0.2.43
[00:02:37]  Downloading time v0.1.40
---
tidy check
[00:04:44] * 555 error codes
[00:04:44] * highest error code: E0712
[00:04:44] * 231 features
[00:04:45] invalid source: "git+https://github.com/Xanewok/rls-data?branch=compilation-options#0b9dbe5bf766a20009def9acac2f997194823373"
[00:04:45] some tidy checks failed
[00:04:45] 
[00:04:45] 
[00:04:45] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:45] 
[00:04:45] 
[00:04:45] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:45] Build completed unsuccessfully in 0:00:49
[00:04:45] Build completed unsuccessfully in 0:00:49
[00:04:45] Makefile:79: recipe for target 'tidy' failed
[00:04:45] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:11d44f9d
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:080b6960:start=1537651696283338024,finish=1537651696287764768,duration=4426744
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:06871760
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0f276a50
travis_time:start:0f276a50
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:0795d281
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Xanewok commented:
As discussed, I removed the extra info, such as `test`, and added the `directory` and `output` fields.

This now includes the associated output file for the compilation (e.g. `libfoo-extra_filename.rmeta`), which can be used together with `CrateSource` and `rmeta`/`rlib` fields.

However, I forgot that to make save-analysis data able to refer to each other, that's exactly why crate disambiguators were added to save-analysis in the first place - we can use this information to generate the dependency information, instead.

@nrc do you think we should hold off including the `output` field?

Xanewok commented on src/librustc_codegen_utils/link.rs:
```diff
@@ -97,6 +97,19 @@ pub fn find_crate_name(sess: Option<&Session>,
+pub fn filename_for_metadata(sess: &Session,
+                             crate_name: &str,
+                             outputs: &OutputFilenames) -> PathBuf {
+    let libname = format!("{}{}", crate_name, sess.opts.cg.extra_filename);
```
Changed the formatting while moving to mimick `filename_for_input`

Xanewok commented on src/librustc_save_analysis/dump_visitor.rs:
```diff
@@ -169,6 +171,54 @@ impl<'l, 'tcx: 'l, 'll, O: DumpOutput + 'll> DumpVisitor<'l, 'tcx, 'll, O> {
+                .filter(|(i, _)| !remap_arg_indices.contains(i))
+                .map(|(_, arg)| {
+                    match input {
+                        Input::File(ref path) if path == Path::new(&arg) => {
```
I decided to propagate `Input` so that I can directly compare the argument to the non-mapped version (`Input` seems to be the only place where we can get input file before it's remapped, apart from scanning the cmd) to replace it with the mapped version.

I was also thinking about trying to remap every argument and see if it matches the already remapped `local_crate_source_file` but that seemed like it's considerably more mechanical work and more error-prone, so in the end I decided against it.

Xanewok commented on src/librustc_save_analysis/span_utils.rs:
```diff
@@ -35,14 +35,24 @@ impl<'a> SpanUtils<'a> {
-            _ => path.to_string(),
+    pub fn make_filename_string(&self, file: &SourceFile) -> String {
+        match &file.name {
+            FileName::Real(path) if !file.name_was_remapped => {
```
In practice every file that's going through `SourceMap` will be remapped but here it seemed to be a missing case for not-yet-remapped absolute path so I included it for the sake of completeness.

Xanewok commented on src/librustc_codegen_llvm/back/link.rs:
```diff
@@ -218,15 +219,6 @@ fn preserve_objects_for_their_debuginfo(sess: &Session) -> bool {
     false
 }
 
-fn filename_for_metadata(sess: &Session, crate_name: &str, outputs: &OutputFilenames) -> PathBuf {
```
Moved it so I can only depend on `librustc_codegen_utils` in the save analysis crate (see [here](https://github.com/rust-lang/rust/pull/54356/commits/a18ca5f9475d4b4afad6c561a9d4a28531bc6ce8))

Xanewok commented:
@nrc hopefully this should be ready to review again (once https://github.com/nrc/rls-data/pull/19 lands).
With this new info I managed to successfully recreate the build plan from save-analysis and run it inside RLS in a Buck project :tada: so hopefully this will be enough as-is for now.

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/432525586) ([raw log](https://api.travis-ci.org/v3/job/432525586/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
########################################                                  56.0%
######################################################################    97.7%
######################################################################## 100.0%
[00:03:24] extracting /checkout/obj/build/cache/2018-09-11/cargo-beta-x86_64-unknown-linux-gnu.tar.gz
[00:03:24]     Updating git repository `https://github.com/Xanewok/rls-data`
[00:03:50]  Downloading getopts v0.2.17
[00:03:50]  Downloading time v0.1.40
[00:03:50]  Downloading lazy_static v0.2.11
[00:03:51]  Downloading num_cpus v1.8.0
---
tidy check
[00:06:02] * 556 error codes
[00:06:02] * highest error code: E0713
[00:06:02] * 231 features
[00:06:03] invalid source: "git+https://github.com/Xanewok/rls-data?branch=compilation-options#5d2c63f02db25a7530ddb52fea9d05cac0401355"
[00:06:03] some tidy checks failed
[00:06:03] 
[00:06:03] 
[00:06:03] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:06:03] 
[00:06:03] 
[00:06:03] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:06:03] Build completed unsuccessfully in 0:00:49
[00:06:03] Build completed unsuccessfully in 0:00:49
[00:06:03] Makefile:79: recipe for target 'tidy' failed
[00:06:03] make: *** [tidy] Error 1

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:24ddf5ad
$ date && (curl -fs --head https://google.com | grep ^Date: | sed 's/Date: //g' || true)
---
travis_time:end:01ab35fc:start=1537805268067661134,finish=1537805268072267080,duration=4605946
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:12b7da6a
$ ln -s . checkout && for CORE in obj/cores/core.*; do EXE=$(echo $CORE | sed 's|obj/cores/core\.[0-9]*\.!checkout!\(.*\)|\1|;y|!|/|'); if [ -f "$EXE" ]; then printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" "$CORE"; gdb --batch -q -c "$CORE" "$EXE" -iex 'set auto-load off' -iex 'dir src/' -iex 'set sysroot .' -ex bt -ex q; echo travis_fold":"end:crashlog; fi; done || true
travis_fold:end:after_failure.4
travis_fold:start:after_failure.5
travis_time:start:0d6f1ba4
travis_time:start:0d6f1ba4
$ cat ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers || true
cat: ./obj/build/x86_64-unknown-linux-gnu/native/asan/build/lib/asan/clang_rt.asan-dynamic-i386.vers: No such file or directory
travis_fold:end:after_failure.5
travis_fold:start:after_failure.6
travis_time:start:335b41e4
$ dmesg | grep -i kill
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

nrc commented on src/Cargo.toml:
```diff
@@ -57,6 +57,7 @@ cargo = { path = "tools/cargo" }
 # that we're shipping as well (to ensure that the rustfmt in RLS and the
 # `rustfmt` executable are the same exact version).
 rustfmt-nightly = { path = "tools/rustfmt" }
+rls-data = { git = "https://github.com/Xanewok/rls-data", branch = "compilation-options" }
```
can use 0.18.1 now

nrc commented:
> One thing I was wondering was the current directory - should we include it in the save-analysis for the invocation or maybe we can assume

I would include it some how, the less we rely on users 'doing the right thing', the better

nrc commented on src/librustc_save_analysis/lib.rs:
```diff
@@ -110,6 +112,24 @@ impl<'l, 'tcx: 'l> SaveContext<'l, 'tcx> {
     }
 
+    // Returns path to the compilation output (e.g. libfoo-12345678.rmeta)
+    pub fn compilation_output(&self, crate_name: &str) -> PathBuf {
```
I thought you said you were removing the output stuff? (Which seems like a good idea to me).

Xanewok replied:
So this doesn't use the file-based approach to figure out dependencies (crate disambiguators are used instead) but if we decide to somehow use save-analysis to lower this info further into low level file-based dep graph (potentially for Cargo build system integration? or maybe we shouldn't use save-analysis for that?) then we need some kind of related file input/output.

I removed `CrateSource` because one-of-dylib-rlib-rmeta-is-some format is clunky to work with but left this for the reasons explained above. Should I remove this as well?

Xanewok commented:
> I would include it some how, the less we rely on users 'doing the right thing', the better

This is included now ([here](https://github.com/rust-lang/rust/pull/54356/files#diff-426e1c382fc57abf697fdeee18e03ddeR213)).

@nrc bumped rls-data to 0.18.1.

nrc commented:
@bors: r+

bors commented:
:pushpin: Commit e5e0d8cfdc452c30b273ab716ac14c9f33fde359 has been approved by `nrc`

<!-- @bors r=nrc e5e0d8cfdc452c30b273ab716ac14c9f33fde359 -->

bors commented:
:lock: Merge conflict

This pull request and the master branch diverged in a way that cannot be automatically merged. Please rebase on top of the latest master branch, and let the reviewer approve again.

<details><summary>How do I rebase?</summary>

Assuming `self` is your fork and `upstream` is this repository, you can resolve the conflict following these steps:

1. `git checkout save-analysis-invocation` *(switch to your branch)*
2. `git fetch upstream master` *(retrieve the latest master)*
3. `git rebase upstream/master -p` *(rebase on top of it)*
4. Follow the on-screen instruction to resolve conflicts (check `git status` if you got lost).
5. `git push self save-analysis-invocation --force-with-lease` *(update this PR)*

You may also read [*Git Rebasing to Resolve Conflicts* by Drew Blessing](http://blessing.io/git/git-rebase/open-source/2015/08/23/git-rebasing-to-resolve-conflicts.html) for a short tutorial.

Please avoid the ["**Resolve conflicts**" button](https://help.github.com/articles/resolving-a-merge-conflict-on-github/) on GitHub. It uses `git merge` instead of `git rebase` which makes the PR commit history more difficult to read.

Sometimes step 4 will complete without asking for resolution. This is usually due to difference between how `Cargo.lock` conflict is handled during merge and rebase. This is normal, and you should still perform step 5 to update this PR.

</details>

<details><summary>Error message</summary>

```text
warning: Cannot merge binary files: src/Cargo.lock (HEAD vs. heads/homu-tmp)
Auto-merging src/librustc_save_analysis/Cargo.toml
Auto-merging src/librustc_driver/lib.rs
Auto-merging src/Cargo.lock
CONFLICT (content): Merge conflict in src/Cargo.lock
Automatic merge failed; fix conflicts and then commit the result.

```

</details>

Xanewok commented:
@nrc rebased and hopefully works now.

nrc commented:
@bors: r+

bors commented:
:hourglass: Testing commit 58a86213abe70ffda0b2edba475d96b3aed95844 with merge 80e6e3e582a9eb9115900d51ae6edecf46d81f6e...
